### PR TITLE
prov/gni: Add ability to grow the int tx pool

### LINKED ISSUE #1199: Crash in GASNet testalign running large number of asynchronous RMA read requests.
Hi there,

I am testing out GASNet using the gni provider and I am running into this issue. I am running GASNet's testalign, which has some sections that issue a large number of read/write RMA ops at once (default: 1000) at different alignments. I am seeing a crash on the part of the test that executes a large number of rma read requests. The call stack is below:

[0] #6  <signal handler called>
[0] #7  0x00002aaaab733875 in raise () from /lib64/libc.so.6
[0] #8  0x00002aaaab734e51 in abort () from /lib64/libc.so.6
[0] #9  0x00002aaaaad53a09 in _gnix_rma_post_rdma_chain_req () from /global/homes/e/epaulson/lib/libfabric.so.1
[0] #10 0x00002aaaaad67abe in _gnix_vc_queue_tx_req () from /global/homes/e/epaulson/lib/libfabric.so.1
[0] #11 0x00002aaaaad57e61 in _gnix_rma () from /global/homes/e/epaulson/lib/libfabric.so.1
[0] #12 0x00002aaaaad1bc49 in gnix_ep_read () from /global/homes/e/epaulson/lib/libfabric.so.1
[0] #13 0x000000000045112b in fi_read (ep=0x8ae8b0, buf=0x94b000, len=32768, desc=0x0, src_addr=3, addr=46915105341441, key=56498863703206, context=0x940b78) at /global/homes/e/epaulson/include/rdma/fi_rma.h:101
[0] #14 0x0000000000459334 in gasnetc_rdma_get (dest_addr=0x94b000, dest=1, src_addr=0x2aab46304001, nbytes=32768, ctxt_ptr=0x940b78) at /global/u2/e/epaulson/gasnet/ofi-conduit/gasnet_ofi.c:1237
[0] #15 0x000000000043aa76 in gasnete_get_nbi_bulk (dest=0x94b000, node=1, src=0x2aab46304001, nbytes=32768) at /global/u2/e/epaulson/gasnet/ofi-conduit/gasnet_extended.c:510
[0] #16 0x000000000040b127 in _gasnet_get_nbi_bulk (dest=0x94b000, node=1, src=0x2aab46304001, nbytes=32768) at /global/u2/e/epaulson/gasnet/extended-ref/gasnet_extended.h:355
[0] #17 0x0000000000412d34 in oneway_nbi_test (iters=1000, nbytes=32768, alignment=1) at /global/u2/e/epaulson/gasnet/tests/testalign.c:184
[0] #18 0x0000000000413d49 in main (argc=1, argv=0x7fffffff6938) at /global/u2/e/epaulson/gasnet/tests/testalign.c:361

Poking around in the GNI code, I see that this from an assert being thrown in this chunk of code:
 if (req->int_tx_buf_e == NULL) {
     req->int_tx_buf_e = _gnix_ep_get_int_tx_buf(ep);
     if (req->int_tx_buf_e == NULL) {
         GNIX_FATAL(FI_LOG_EP_DATA, "RAN OUT OF INT_TX_BUFS");
         return -FI_ENOSPC;
         /* TODO Create growable list of buffers */
     }
 }

To me it seems like instead of an assertion, this should bubble back up to the caller in the form of -FI_EAGAIN so they know to poll to free up some of these buffers. Although it does appear that this code path is being actively worked on based off of the TODO.

This is only occurring on the head of the repository, not on the 1.4 release.

### PR DESCRIPTION
**Pull request by chuckfossen**
fixes ofi-cray/libfabric-cray#1199

Signed-off-by: Chuck Fossen <chuckf@cray.com>

### CODE REVIEW

**Comment by hppritcha** [comment]
LANL-CI is having trouble getting allocations today on edison.

File: prov/gni/src/gnix_ep.c [hunk 1]
```diff
@@ -66,79 +66,137 @@
...
+	struct gnix_fid_mem_desc *md = NULL;
+	struct gnix_int_tx_buf *tx_buf_list;
+	struct gnix_int_tx_ptrs *tx_ptrs;
 
 	assert(ep);
 
-	int_tx_bufs = malloc(GNIX_INT_TX_POOL_SIZE * GNIX_INT_TX_BUF_SZ);
-	if (int_tx_bufs == NULL) {
-		GNIX_FATAL(FI_LOG_EP_DATA, "\n");
+	if (ep->int_tx_pool.nbufs >= GNIX_INT_TX_POOL_COUNT)
```
Comment (sungeunchoi) [inline prov/gni/src/gnix_ep.c]: Wondering if there should be a warning here.  Or maybe a GNIX_INFO?

File: prov/gni/src/gnix_ep.c [hunk 2]
```diff
@@ -66,79 +66,137 @@
...
-		GNIX_DEBUG(FI_LOG_EP_DATA, "fi_close returned: %s\n", fi_strerror(-ret));
-	}
+		ret = fi_close(&tx_ptrs->md->mr_fid.fid);
 
-	if (ep->int_tx_pool.buf_ptr != NULL) {
-		free(ep->int_tx_pool.buf_ptr);
-		ep->int_tx_pool.buf_ptr = NULL;
-	}
+		if (ret != FI_SUCCESS) {
+			GNIX_DEBUG(FI_LOG_EP_DATA, "fi_close returned: %s\n",
```
Comment (sungeunchoi) [inline prov/gni/src/gnix_ep.c]: Maybe this should this be GNIX_WARN with a little more detail?  What would be the consequences if this failed?

File: prov/gni/src/gnix_ep.c [hunk 3]
```diff
@@ -66,79 +66,137 @@
...
 void _gnix_ep_int_tx_pool_fini(struct gnix_fid_ep *ep)
 {
 	int ret;
+	struct slist_entry *e;
+	struct gnix_int_tx_ptrs *tx_ptrs;
 
 	assert(ep);
 
 	if (ep->int_tx_pool.enabled == false)
 		return;
```
Comment (sungeunchoi) [inline prov/gni/src/gnix_ep.c:178]: I think we should grab the lock here and release it after we set enabled to false.

File: prov/gni/src/gnix_ep.c (hunk 1 above)
Comment (hppritcha) [reply prov/gni/src/gnix_ep.c]: I'd go for warning for now

**Review by hppritcha — APPROVED** [review]

**Comment by hppritcha** [comment]
bot:lanl:retest

File: prov/gni/src/gnix_ep.c (hunk 3 above)
Comment (chuckfossen) [reply prov/gni/src/gnix_ep.c:178]: yes, I agree.

File: prov/gni/src/gnix_ep.c (hunk 2 above)
Comment (chuckfossen) [reply prov/gni/src/gnix_ep.c]: this debug was already there. I can change to warn but I'm not sure of the consequences of the deregistration failing.

**Comment by hppritcha** [comment]
okay I think things are cleared up with AWS jenkins:
bot:lanl:retest

**Review by sungeunchoi — APPROVED** [review]

**Comment by sungeunchoi** [comment]
@hppritcha Just to clarify, are you planning to add to this PR or should it be merged?

**Comment by hppritcha** [comment]
should be merged.  I'll open the subsequent PR later this week.

**Pull request merged**